### DISCUSSION TITLE
Initial Config Struggle

### DISCUSSION THREAD

**Discussion by lamtech58** [arkivanov/Decompose#786] [Q&A]
My first attemp to initialize the root component while waiting that the initialConfig is initialized first, it is working but feeling this way is not good at all. Please provide me with a best solution

`private val root = MutableStateFlow<RootComponent?>(null)
    private val initialConfig = MutableStateFlow<Root.Configuration?>(null)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        combine(
            dataStore.data,
            paymentDataStore.data,
            localAppDataStore.data,
            locationDataStore.data
        ){ mAppConfig, mPaymentState, mLocalAppData, mUserLocation ->

            
            initialConfig.update {
                if(mLocalAppData.isFirstAppLaunch) Root.Configuration.OnBoardingScreen else Root.Configuration.ClientHomeScreen
            }

            appConfig.update { mAppConfig }
            paymentState.update { mPaymentState }
            localAppData.update { mLocalAppData }
            userLocation.update { mUserLocation }

            if(mUserLocation.isTheCurrentAddressConfirmed){
                fusedLocationClient.removeLocationUpdates(locationCallback)
            }
        }.stateIn(lifecycleScope, SharingStarted.Eagerly, null)

        lifecycleScope.launch {
                initialConfig.collectLatest { initialConfig ->
                    if(initialConfig != null){
                        root.update {
                            retainedComponent { componentContext ->
                                RootComponent(
                                    componentContext = componentContext,
                                    context = applicationContext,
                                    initialConfiguration = initialConfig,
                                    onToast = { uiText ->
                                        toaster.update { uiText.asString(applicationContext) }
                                    },
                                    signIn = ::signIn
                                )
                            }
                        }
                    }
            }
        }

        installSplashScreen()
            .setKeepOnScreenCondition {
                root.value == null
            }

        setContent {

            val mRoot by root.collectAsStateWithLifecycle()

            MyHalalTheme {
                if(mRoot != null) {
                    AppEntry(
                        root = mRoot!!,
                        onSignOut = {},
                        visitMyNaamaFacebookPage = ::visitMyNaamaFacebookPage,
                        writeToMyNaamaViaWhatsapp = ::sendWhatsappMessage,
                        sendMyNaamaAnEmail = ::sendMyNaamaAnEmail
                    )
                }
            }
        }

    }`

**Comment by arkivanov** [comment]
Please join the related discussion: #778.
